hail/src/main/scala/is/hail/variant/MatrixTable.scala

```diff
@@ -1525,100 +1529,103 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) {
-        val inAdjustment = it.hasNext
-        val rvIterator = f(newContext)
+      val region = context.region
+      val rv2 = RegionValue()
+      val rvb = new RegionValueBuilder()
 
-        val region = context.region
-        val rv2 = RegionValue()
-        val rvb = new RegionValueBuilder()
-        rvIterator.flatMap { rv =>
-          val locus = UnsafeRow.readLocus(rv.region, localRVRowType.loadField(rv, locusIndex), rg)
+      def pushRow(locus: Locus, row: RegionValue) {
+        val cpRegion = fetchRegion()
+        rvb.set(cpRegion)
+        rvb.clear()
+        rvb.start(localRVRowType)
+        rvb.startStruct()
+        rvb.addAllFields(localRVRowType, row)
+        rvb.endStruct()
+        deque.push(locus -> RegionValue(cpRegion, rvb.end()))
+      }
 
-          val result = if (inAdjustment) {
-            None
-          } else {
-            def discard(x: (Locus, RegionValue)): Boolean = x != null && (x._1.position < locus.position - basePairs
-              || x._1.contig != locus.contig)
+      def getLocus(row: RegionValue): Locus =
+        UnsafeRow.readLocus(row.region, localRVRowType.loadField(row, locusIndex), rg)
 
-            while (discard(deque.peekLast()))
-              recycleRegion(deque.removeLast()._2.region)
+      val unsafeRow = new UnsafeRow(localRVRowType)
+      val keyView = new KeyedRow(unsafeRow, localKeyFieldIdx)
+      while (bit.hasNext && { unsafeRow.set(bit.head); rangeBoundsBc.value(i).isAbovePosition(keyOrdering, keyView) }) {
+        val rv = bit.next()
+        pushRow(getLocus(rv), rv)
+      }
 
-            val rvs = deque.iterator().asScala.map(_._2).toArray
+      bit.map { rv =>
+        val locus = UnsafeRow.readLocus(rv.region, localRVRowType.loadField(rv, locusIndex), rg)
 
-            rvb.set(region)
-            rvb.clear()
-            region.clear()
-            rvb.start(newType.rvRowType)
-            rvb.startStruct()
-            rvb.addFields(localRVRowType, rv, nonEntryIndices)
+        def discard(x: (Locus, RegionValue)): Boolean = x != null && (x._1.position < locus.position - basePairs
+          || x._1.contig != locus.contig)
 
-            // prev_rows
-            rvb.startArray(rvs.length)
-            var j = 0
-            while (j < rvs.length) {
-              val rvj = rvs(j)
-              rvb.startStruct()
-              rvb.addFields(localRVRowType, rvj, nonEntryIndices)
-              rvb.endStruct()
-              j += 1
-            }
-            rvb.endArray()
+        while (discard(deque.peekLast()))
+          recycleRegion(deque.removeLast()._2.region)
 
-            rvb.startArray(nCols)
+        val rvs = deque.iterator().asScala.map(_._2).toArray
 
-            val entriesOffset = localRVRowType.loadField(rv, entriesIndex)
-            val prevEntriesOffsets = rvs.map(localRVRowType.loadField(_, entriesIndex))
+        rvb.set(region)
+        rvb.clear()
+        region.clear()
```

Comment:
erm. I know this was here before, but I feel like you're not allowed to clear the region given to you.

Reply:
I think you're right. (If you're not certain, then nobody is, which is concerning.)